### PR TITLE
[RWRoute] Versal optimizations

### PR DESCRIPTION
**Pull request by WenhaoLin-AMD**
This PR:
- Applies https://github.com/Xilinx/RapidWright/pull/1095 to Versal nodes based on their intent codes -- ensuring that NON_LOCAL -> LOCAL node transition only happens in the vicinity of the sink tile
  - With special handling for `NODE_(CLE|INTF)_[BC]NODE` nodes to prevent them from being used to travel between `INT` tiles, but still used to "bounce" (back) into the original (sink) tile
- Further applies https://github.com/Xilinx/RapidWright/pull/1098 to Versal nodes based on their wire or tile name to ensure that only the relevant east/west side of the INT tile is explored
  - Note that `NODE_CLE_[BC]NODE` nodes can reach `NODE_INODE`s on the opposite side of the former's wire name
- Add a `RouteNodeType.LOCAL_RESERVED` as part of a Versal-only *heuristic* to reserve `[BC]NODE`s uphill of a `NODE_(CLE|INTF)_CTRL` sink, since `[BC]NODE`s (especially CNODEs) are the only way to reach CTRL sinks, while also providing extra connectivity to `NODE_INODE`s in the `INT` tile.
  - e.g. On 1c2b471, for `finn_radioml` using `LOCAL_RESERVED` requires 119M pops over 10 iterations to route, but without needed 176M pops over 17 iterations.
  - The downside to this heuristic is that all CNODEs are reserved for even one CTRL sink, meaning that just one sink will prevent any other nets from using those CNODEs

Results:
```
			Before (a3af7dc)		After (1c2b471)	
			Nodes popped	Runtime (s)	Nodes popped	Runtime (s)
boom_med_pb		  308,212,121	 201		 92,043,436	 79
boom_soc		3,320,844,555	2184		831,951,144	698
corescore_500		  381,326,217	 249		141,432,563	133
corescore_500_pb	  393,285,727	 260		147,020,615	137
finn_radioml		  401,124,667	 253		118,592,148	101
koios_dla_like_large	1,329,175,620	 923		270,148,933	263
logicnets_jscl		   85,632,115	  58		 42,305,715	 33
vtr_lu64peeng		  429,006,483	 295		194,564,833	175
vtr_mcml		  185,015,595	 127		 79,474,273	 71
```

Overall, 2.7x improvement in wall clock RWRoute runtime, 3.6x the number of nodes popped.

### CODE REVIEW

**Comment by eddieh-xlnx** [comment]
This PR:
- Applies https://github.com/Xilinx/RapidWright/pull/1095 to Versal nodes based on their intent codes -- ensuring that NON_LOCAL -> LOCAL node transition only happens in the vicinity of the sink tile
  - With special handling for `NODE_(CLE|INTF)_[BC]NODE` nodes
- Further applies https://github.com/Xilinx/RapidWright/pull/1098 to Versal nodes based on their wire or tile name to ensure that only the relevant east/west side of the INT tile is explored
  - Note that `NODE_CLE_[BC]NODE` nodes can reach `NODE_INODE`s on the opposite side of the former's wire name
- Add a `RouteNodeType.LOCAL_RESERVED` as part of a Versal-only heuristic to reserve `[BC]NODE`s uphill of a `NODE_(CLE|INTF)_CTRL` sink, since `[BC]NODE`s (especially CNODEs) are the only way to reach CTRL sinks, while also providing extra connectivity to `NODE_INODEs` in the `INT` tile.

File: src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java [hunk 1]
```diff
@@ -175,69 +182,115 @@ protected RouteNodeGraph(Design design, RWRouteConfig config, Map<Tile, RouteNod
...
-                continue;
-            }
+            } else {
+                assert(isVersal);
 
-            if (eastWestPattern == null) {
-                // FIXME:
-                assert(series == Series.Versal);
-                continue;
+                if (!EnumSet.of(IntentCode.NODE_IMUX, IntentCode.NODE_PINBOUNCE, IntentCode.NODE_INODE,
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java:224]: Would it make sense to statically construct this set?

File: src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java [hunk 2]
```diff
@@ -175,69 +182,115 @@ protected RouteNodeGraph(Design design, RWRouteConfig config, Map<Tile, RouteNod
...
+                        k -> new BitSet[]{new BitSet(), new BitSet()});
+                BitSet eastWires = eastWestWires[0];
+                BitSet westWires = eastWestWires[1];
+                for (int wireIndex = 0; wireIndex < intfTile.getWireCount(); wireIndex++) {
+                    IntentCode baseIntentCode = intfTile.getWireIntentCode(wireIndex);
+                    if (baseIntentCode != IntentCode.NODE_INTF_BNODE && baseIntentCode != IntentCode.NODE_INTF_CNODE) {
+                        continue;
+                    }
+
+                    if (EnumSet.of(TileTypeEnum.INTF_LOCF_TR_TILE,
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java:285]: Also move out to static instantiation?  Looks like its in a double-nested for loop.

File: src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java (hunk 1 above)
Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java]: I've hoisted it outside of the for loop.

Given that this set is only used inside the constructor, and the `RouteNodeGraph` object is only instantiated once per RWRoute invocation, I'd prefer the code readability benefits over making it a static object in the class body.

**Review by clavin-xlnx — APPROVED** [review]

**Pull request merged**